### PR TITLE
feature: Add csharp code array_deque and linkedlist_deque

### PR DESCRIPTION
**Pull request by hpstory**
If this PR is related to coding or code translation, please read the [contribution guideline](https://github.com/krahets/hello-algo/issues/15#issue-1464069591), fill out the checklist, and paste the console outputs to the PR.

- [x] I've tested the code and ensured the outputs are the same as the outputs of reference codes.
- [x] I've checked the codes (formatting, comments, indentation, file header, etc) carefully.
- [x] The code does not rely on a particular environment or IDE and can be executed on a standard system (Win, macOS, Ubuntu).

linkedlist_deque测试结果
linkedlist_deque

array_deque测试结果
array_deque

### CODE REVIEW

File: codes/csharp/chapter_stack_and_queue/array_deque.cs [hunk 1]
```diff
@@ -0,0 +1,173 @@
+﻿using NUnit.Framework;
```
Comment (krahets) [inline codes/csharp/chapter_stack_and_queue/array_deque.cs:1]: Please add the file header

File: codes/csharp/chapter_stack_and_queue/array_deque.cs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+        /* 队首入队 */
+        public void pushFirst(int num)
+        {
+            if (queSize == capacity())
+            {
+                Console.WriteLine("双向队列已满");
+                return;
+            }
+
```
Comment (krahets) [inline codes/csharp/chapter_stack_and_queue/array_deque.cs:54]: Please remove this empty line

File: codes/csharp/chapter_stack_and_queue/array_deque.cs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        /* 队尾入队 */
+        public void pushLast(int num)
+        {
+            if (queSize == capacity())
+            {
+                Console.WriteLine("双向队列已满");
+                return;
+            }
+
```
Comment (krahets) [inline codes/csharp/chapter_stack_and_queue/array_deque.cs:71]: Please remove this empty line

File: codes/csharp/chapter_stack_and_queue/array_deque.cs [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        }
+
+        /* 访问队首元素 */
+        public int peekFirst()
+        {
+            if (isEmpty())
+            {
+                throw new InvalidOperationException();
+            }
+
```
Comment (krahets) [inline codes/csharp/chapter_stack_and_queue/array_deque.cs:104]: Same as above

File: codes/csharp/chapter_stack_and_queue/array_deque.cs [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        }
+
+        /* 访问队尾元素 */
+        public int peekLast()
+        {
+            if (isEmpty())
+            {
+                throw new InvalidOperationException();
+            }
+
```
Comment (krahets) [inline codes/csharp/chapter_stack_and_queue/array_deque.cs:115]: Same as above

File: codes/csharp/chapter_stack_and_queue/linkedlist_deque.cs [hunk 6]
```diff
@@ -0,0 +1,207 @@
+﻿using NUnit.Framework;
```
Comment (krahets) [inline codes/csharp/chapter_stack_and_queue/linkedlist_deque.cs:1]: Please add the file header

File: codes/csharp/chapter_stack_and_queue/linkedlist_deque.cs [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+                rear = node;
+            }
+            // 队首入队操作
+            else if (isFront)
+            {
+                // 将 node 添加至链表头部
+                front.prev = node;
+                node.next = front;
+                front = node; // 更新头结点
+                              // 队尾入队操作
```
Comment (krahets) [inline codes/csharp/chapter_stack_and_queue/linkedlist_deque.cs:61]: Incorrect position

File: codes/csharp/chapter_stack_and_queue/linkedlist_deque.cs [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+                                 // 删除头结点
+                ListNode fNext = front.next;
+                if (fNext != null)
+                {
+                    fNext.prev = null;
+                    front.next = null;
+                }
+
+                front = fNext;   // 更新头结点
+                                 // 队尾出队操作
```
Comment (krahets) [inline codes/csharp/chapter_stack_and_queue/linkedlist_deque.cs:109]: Incorrect position

**Review by krahets — COMMENTED** [review]
Please address the comments, thanks!

File: codes/csharp/chapter_stack_and_queue/array_deque.cs (hunk 1 above)
Comment (hpstory) [reply codes/csharp/chapter_stack_and_queue/array_deque.cs]: Fixed

File: codes/csharp/chapter_stack_and_queue/array_deque.cs (hunk 2 above)
Comment (hpstory) [reply codes/csharp/chapter_stack_and_queue/array_deque.cs]: Fixed

File: codes/csharp/chapter_stack_and_queue/array_deque.cs (hunk 3 above)
Comment (hpstory) [reply codes/csharp/chapter_stack_and_queue/array_deque.cs]: Fixed

File: codes/csharp/chapter_stack_and_queue/array_deque.cs (hunk 4 above)
Comment (hpstory) [reply codes/csharp/chapter_stack_and_queue/array_deque.cs]: Fixed

File: codes/csharp/chapter_stack_and_queue/array_deque.cs (hunk 5 above)
Comment (hpstory) [reply codes/csharp/chapter_stack_and_queue/array_deque.cs]: Fixed

File: codes/csharp/chapter_stack_and_queue/linkedlist_deque.cs (hunk 6 above)
Comment (hpstory) [reply codes/csharp/chapter_stack_and_queue/linkedlist_deque.cs]: Fixed

File: codes/csharp/chapter_stack_and_queue/linkedlist_deque.cs (hunk 7 above)
Comment (hpstory) [reply codes/csharp/chapter_stack_and_queue/linkedlist_deque.cs]: Fixed

File: codes/csharp/chapter_stack_and_queue/linkedlist_deque.cs (hunk 8 above)
Comment (hpstory) [reply codes/csharp/chapter_stack_and_queue/linkedlist_deque.cs]: Fixed

**Comment by hpstory** [comment]
> Please address the comments

Thanks for the review, and sorry for the code style issue.
My C# code is based on StyleCop Analyzers which need to add a blank line before a single line comments or closing brace

Reference: 
[SA1513 A closing brace within a C# element, statement, or expression is not followed by a blank line.](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1513.md)
[SA1515 A single-line comment within C# code is not preceded by a blank line.](https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/SA1515.md)

**Review by krahets — APPROVED** [review]
Thanks!

**Pull request merged**